### PR TITLE
API-114: create repositories for the API

### PR DESCRIPTION
**Pull request by skeleton**
[//]: <> (<3 Thanks for taking the time to contribute! You're awesome! <3)

Implements our own repositories for the API.

[//]: <> (If you've never contributed to this repository before, please read https://github.com/akeneo/pim-community-dev/blob/master/.github/CONTRIBUTING.md)

**Description (for Contributor and Core Developer)**

[//]: <> (What does this Pull Request do? reference the related issue?)

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added Behats                      | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

**Comment by phaseinducer** [comment]
I reformulate what I said IRL:
- For the products: add count and pagination capabilities to the PQB.
- For other entities: create new repos implementing a single interface.

**Comment by jjanvier** [comment]
all the following is under the "imho" clause

1. about adding methods to the PQB:
- If I'm right, the idea of the PQB was to be able to get all results regarding a particular search (which works perfectly for the grid we have, of for an infinite scroll we could have). That's why we choose to implement it via a cursor. If we keep this approach, the methods `setMaxResult` and `setFirstResult` does not make sense for the PQB. For me, the PQB _is not_ a facade to a product Doctrine repository.
- about the `count`. Except on your API use case, I can't imagine another use case where you need to count the number of results without applying the query. I would say the API is really a specific use case, and I wouldn't put it in the PQB.
2. if everything is _exactly_ the same for other entities, yeah, use a generic class ok. Maybe at least different services so that there is no BC break in the future.

To sum up, I'm in favor of this PR. This goes towards the direction of decoupling features (see my long answer here https://github.com/akeneo/pim-community-dev/pull/5569#discussion_r97776705). 

I find this simple, and let the door open to future possibilities.

File: src/Pim/Component/Api/Repository/ApiResourceRepositoryInterface.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace Pim\Component\Api\Repository;
+
+use Akeneo\Component\StorageUtils\Repository\IdentifiableObjectRepositoryInterface;
+
+/**
+ * Repository interface for simple API resources
```
Comment (phaseinducer) [inline src/Pim/Component/Api/Repository/ApiResourceRepositoryInterface.php]: Maybe explain what is the purpose of this interface? Why it extends the two others, explain that it's generic, suggest that the implementation should decorate other repos, etc.

File: src/Pim/Bundle/ApiBundle/Doctrine/ORM/Repository/AttributeRepository.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            }
+
+            return (int) $qb
+                ->select('COUNT(DISTINCT o.id)')
+                ->getQuery()
+                ->getSingleScalarResult();
+        } catch (UnexpectedResultException $e) {
+            return 0;
+        }
+    }
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Doctrine/ORM/Repository/AttributeRepository.php]: what's the difference with ApiResourceRepository ?

Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Doctrine/ORM/Repository/AttributeRepository.php]: method getIdentifierCode() is added

File: src/Pim/Bundle/ApiBundle/Doctrine/MongoDBODM/Repository/ProductRepository.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        return $this->productRepository->findOneByIdentifier($identifier);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function searchAfterOffset(ProductQueryBuilderInterface $pqb, $limit, $offset)
+    {
+        $qb = $pqb->getQueryBuilder();
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Doctrine/MongoDBODM/Repository/ProductRepository.php]: you modify the state of the pqb. You should clone it.

File: src/Pim/Bundle/ApiBundle/Doctrine/MongoDBODM/Repository/ProductRepository.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            ->getQuery()
+            ->execute();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function count(ProductQueryBuilderInterface $pqb)
+    {
+        return (int) $pqb->getQueryBuilder()->select('_id')
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Doctrine/MongoDBODM/Repository/ProductRepository.php]: you modify the state of the pqb too. Is your integration test ok with mongo ? Because it should impact the call of the searchAfterOffset (which use the same pqb).

File: src/Pim/Bundle/ApiBundle/Doctrine/ORM/Repository/ProductRepository.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    {
+        return $this->productRepository->findOneByIdentifier($identifier);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function searchAfterOffset(ProductQueryBuilderInterface $pqb, $limit, $offset)
+    {
+        $qb = $pqb->getQueryBuilder();
```
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Doctrine/ORM/Repository/ProductRepository.php]: you modify the state of the pqb.

File: src/Pim/Bundle/ApiBundle/Doctrine/MongoDBODM/Repository/ProductRepository.php (hunk 4 above)
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Doctrine/MongoDBODM/Repository/ProductRepository.php]: distinct is not needed here as there is never join with other entities, so no risk to have id duplicated.

File: src/Pim/Bundle/ApiBundle/Doctrine/ORM/Repository/ProductRepository.php (hunk 5 above)
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Doctrine/ORM/Repository/ProductRepository.php]: yep, I added on count() method but forgot it on this method. thank
